### PR TITLE
login route

### PR DESCRIPTION
**Pull request by NourAdawi**

### CODE REVIEW

**Comment by NourAdawi** [comment]
@Mu7ammadAbed

**Review by rawand1998 — APPROVED** [review]

**Review by mohammedsalah7 — DISMISSED** [review]

File: server/controllers/logoutController.js [hunk 1]
```diff
@@ -0,0 +1,4 @@
+exports.logoutRouter = (req, res) => {
+  res.clearCookie('authToken');
+  res.redirect('/');
```
Comment (Mu7ammadAbed) [inline server/controllers/logoutController.js:3]: You shouldn't respond with a redirect from the backend, you have to send a response with a prober message in order to handle it correctly at the backend
```suggestion
  res.clearCookie('authToken').json({ statusCode: 200, message: 'Logged out successfully' });
```

File: server/middlewares/privateRoute.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+const jwt = require('jsonwebtoken');
+
+exports.privateRoute = (req, res, next) => {
+  const { authToken } = req.cookies;
+  if (!authToken) {
+    return res.status(401)
+      .redirect('/login');
```
Comment (Mu7ammadAbed) [inline server/middlewares/privateRoute.js:7]: You shouldn't redirect ...

File: server/middlewares/privateRoute.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    req.user = verified;
+    next();
+  } catch (err) {
+    res.status(400).json({ success: false, msg: 'Invalid token!' });
+  }
+};
+
+exports.notLogged = (req, res, next) => {
+  const { authToken } = req.cookies;
+  if (authToken) return res.redirect('/');
```
Comment (Mu7ammadAbed) [inline server/middlewares/privateRoute.js:20]: What is this exactly ?

**Review by muhammadharoun — APPROVED** [review]

File: server/middlewares/privateRoute.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+exports.privateRoute = (req, res, next) => {
+  const { authToken } = req.cookies;
+  if (!authToken) {
+    return res.status(401)
+      .redirect('/login');
+  }
+  try {
+    const verified = jwt.verify(authToken, process.env.TOKEN_SERCRET);
+    req.user = verified;
```
Comment (Mu7ammadAbed) [inline server/middlewares/privateRoute.js:11]: You are making this function work synchronus

File: server/utiles/createToken.js [hunk 5]
```diff
@@ -0,0 +1,11 @@
+const jwt = require('jsonwebtoken');
+
+exports.createToken = (id, username, secret, res) => {
+  const token = jwt.sign({ id, username }, secret);
```
Comment (Mu7ammadAbed) [inline server/utiles/createToken.js:4]:
```suggestion
const { sign } = require('jsonwebtoken');

exports.createToken = (id, username, secret, res) => {
  const token = sign({ id, username }, secret);
```

File: server/controllers/loginController.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+const { createToken } = require('../utiles/createToken');
+
+exports.loginPost = async (req, res) => {
+  // validate data before log in
+  const { error } = await loginValidation(req.body);
+  if (error) return res.status(400).send(error.details[0].message);
+
+  // check if the email exists
+  const user = await emailExists(req.body.email);
+  if (user.rowCount === 0) return res.status(400).json({ success: false, msg: 'Invalid Email, you have to sign up first!' });
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:13]:
```suggestion
  const { rows, rowCount } = await emailExists(email);
  if (rowCount === 0) return res.status(400).json({ success: false, msg: 'Invalid Email, you have to sign up first!' });
```

File: server/controllers/loginController.js [hunk 7]
```diff
@@ -0,0 +1,28 @@
+const bcrypt = require('bcrypt');
+const emailExists = require('../database/queries/checkEmailExists');
+const { loginValidation } = require('../utiles/validation/logInValidation');
+const { createToken } = require('../utiles/createToken');
+
+exports.loginPost = async (req, res) => {
+  // validate data before log in
+  const { error } = await loginValidation(req.body);
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:8]:
```suggestion
  const { error, value: { email, password } } = await loginValidation(req.body);
```

File: server/controllers/loginController.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+  if (error) return res.status(400).send(error.details[0].message);
+
+  // check if the email exists
+  const user = await emailExists(req.body.email);
+  if (user.rowCount === 0) return res.status(400).json({ success: false, msg: 'Invalid Email, you have to sign up first!' });
+
+  // check password
+  const dbPassword = user.rows[0].password;
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:16]:
```suggestion
  const { dbPassword: password } = rows[0];
```

File: server/controllers/loginController.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+  const validPassword = await bcrypt.compare(req.body.password, dbPassword);
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:17]:
```suggestion
  const validPassword = await bcrypt.compare(password, dbPassword);
```

File: server/controllers/loginController.js [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  if (!validPassword) return res.status(400).json({ success: false, msg: 'Incorrect password' });
+
+  // create token and store it in cookies
+  const { id, name } = user.rows[0];
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:21]:
```suggestion
  const { id, name } = rows[0];
```

File: server/controllers/loginController.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+  createToken(id, name, process.env.TOKEN_SERCRET, res);
+  res.send('you are logged in now!');
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:23]: Send a JSON instead in order to handle it correctly at the frontend

File: server/controllers/loginController.js [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+};
+
+exports.loginGet = (req, res) => {
+  res.send('home page');
```
Comment (Mu7ammadAbed) [inline server/controllers/loginController.js:27]: Send a JSON instead in order to handle it correctly at the frontend

**Review by rawand1998 — APPROVED** [review]

**Review by mohammedsalah7 — APPROVED** [review]

**Review by muhammadharoun — APPROVED** [review]

**Pull request merged**